src/orange/orangeinp/detail/CsgUnitBuilder.hh

```diff
@@ -0,0 +1,161 @@
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Label.hh"
+#include "orange/OrangeTypes.hh"
+#include "orange/construct/CsgTypes.hh"
+#include "orange/construct/detail/LocalSurfaceInserter.hh"
+
+#include "CsgUnit.hh"
+
+namespace celeritas
+{
+namespace orangeinp
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Help construct a unit's mutable CSG representation.
+ *
+ * This keeps track of the CSG and surface nodes during construction. It holds
+ * the "construction-time" properties like the local surface inserter. The
+ * input "CSG Unit" must exceed the lifetime of this builder.
+ */
+class CsgUnitBuilder
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Tol = Tolerance<>;
+    using Metadata = CsgUnit::Metadata;
+    using NodeId = CsgUnit::NodeId;
+    using BBox = CsgUnit::BBox;
+    //!@}
+
+  public:
+    // Construct with a "processed unit" to build.
+    inline CsgUnitBuilder(CsgUnit*, Tol const& tol);
+
+    //// ACCESSORS ////
+
+    //! Tolerance, needed for surface simplifier
+    Tol const& tol() const { return tol_; }
+
+    // Access a typed surface, needed for clipping with deduplicated surface
+    template<class S>
+    inline S const& get_surface(NodeId) const;
+
+    //// MUTATORS ////
+
+    // Insert a surface by forwarding to the surface inserter
+    template<class... Args>
+    inline NodeId insert_surface(Args&&... args);
+
+    // Insert a CSG node by forwarding to the CsgTree
+    template<class... Args>
+    inline NodeId insert_csg(Args&&... args);
+
+    //! Insert node metadata
+    inline void insert_md(NodeId node, Metadata&& md);
+
+    //! Set a bounding box for a node
+    void set_bbox(NodeId, BBox const&);
+
+    // Mark a CSG node as a volume of real space
+    LocalVolumeId insert_volume(NodeId);
+
+    // Fill a volume node with a material
+    void fill_volume(LocalVolumeId, MaterialId);
+
+    // Fill a volume node with a daughter
+    void fill_volume(LocalVolumeId, UniverseId, VariantTransform&& vt);
+
+    // Set an exterior node
+    void set_exterior(NodeId);
+
+  private:
+    using LocalSurfaceInserter = ::celeritas::detail::LocalSurfaceInserter;
+
+    CsgUnit* unit_;
+    Tol tol_;
+    LocalSurfaceInserter insert_surface_;
+
+    // Get a variant surface from a node ID
+    VariantSurface const& get_surface_impl(NodeId nid) const;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with a fresh unit.
+ */
+CsgUnitBuilder::CsgUnitBuilder(CsgUnit* u, Tolerance<> const& tol)
+    : unit_{u}, tol_{tol}, insert_surface_{&unit_->surfaces, tol}
+{
+    CELER_EXPECT(unit_);
+    CELER_EXPECT(unit_->metadata.size() == 0 && unit_->bboxes.size() == 0
```

Comment:
Maybe `CsgUnit` could have an `is_fresh` or `is_empty` method to replace this line